### PR TITLE
MINOR: More efficient midpoint calc for AbstractIndex

### PR DESCRIPTION
**Pull request by cmccabe**

### CODE REVIEW

**Comment by cmccabe** [comment]
retest this please

**Comment by cmccabe** [comment]
retest this please

**Review by satishd — COMMENTED** [review]
@cmccabe Thanks for the PR. Can you resolve the changes with latest trunk?

File: core/src/main/scala/kafka/log/AbstractIndex.scala [hunk 1]
```diff
@@ -374,7 +374,7 @@ abstract class AbstractIndex[K, V](@volatile var file: File, val baseOffset: Lon
       var lo = begin
       var hi = end
       while(lo < hi) {
-        val mid = ceil(hi/2.0 + lo/2.0).toInt
+        val mid = lo + (hi - lo) / 2
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/AbstractIndex.scala]: Is there a reason why we can't do the same as Java's binary search? That is:

```java
int mid = (low + high) >>> 1;
```

Comment (ijuma) [reply core/src/main/scala/kafka/log/AbstractIndex.scala]: Explanation in https://ai.googleblog.com/2006/06/extra-extra-read-all-about-it-nearly.html

Comment (satishd) [reply core/src/main/scala/kafka/log/AbstractIndex.scala]: Reference for the above logic is here [DualPivotQuicksort](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/tip/src/share/classes/java/util/DualPivotQuicksort.java#l289)

Comment (ijuma) [reply core/src/main/scala/kafka/log/AbstractIndex.scala]: Bump @cmccabe

**Comment by ijuma** [comment]
@cmccabe I pushed a commit. If that looks good to you, you can merge.

**Review by ijuma — APPROVED** [review]
LGTM

**Comment by cmccabe** [comment]
retest this please

**Comment by ijuma** [comment]
retest this please

**Comment by ijuma** [comment]
retest this please

**Comment by hachikuji** [comment]
This PR causes `LogCleanerTest .testRecoveryAfterCrash` to fail in an infinite loop in `binarySearch` with lo=15 and hi=16.

**Comment by ijuma** [comment]
Fixed via https://github.com/apache/kafka/pull/7702, thanks @hachikuji

**Pull request merged**